ma_helper.c

```diff
@@ -535,40 +573,53 @@ size_t MADB_GetDisplaySize(MYSQL_FIELD *Field, MARIADB_CHARSET_INFO *charset)
   case MYSQL_TYPE_TIMESTAMP:
   case MYSQL_TYPE_DATETIME:
     return SQL_TIMESTAMP_LEN + MADB_FRACTIONAL_PART(Field->decimals);
+  // LONGBLOB/LONGTEXT - like types, length is max size in both bytes and characters
+  case MYSQL_TYPE_TINY_BLOB:
   case MYSQL_TYPE_BLOB:
-  case MYSQL_TYPE_ENUM:
-  case MYSQL_TYPE_GEOMETRY:
-  case MYSQL_TYPE_LONG_BLOB:
   case MYSQL_TYPE_MEDIUM_BLOB:
-  case MYSQL_TYPE_SET:
-  case MYSQL_TYPE_STRING:
-  case MYSQL_TYPE_TINY_BLOB:
-  case MYSQL_TYPE_VARCHAR:
+  case MYSQL_TYPE_LONG_BLOB:
+    return MIN(Field->length, S2_1_GB);
+  // VARCHAR(n), VARBINARY(n), ENUM, SET
   case MYSQL_TYPE_VAR_STRING:
+  // CHAR(n), BINARY(n)
+  case MYSQL_TYPE_STRING:
   {
     if (MADB_FIELD_IS_BINARY(Field))
     {
-      return Field->length*2; /* ODBC specs says we should give 2 characters per byte to display binaray data in hex form */
+      return MIN(Field->length, S2_1_GB) * 2; /* ODBC specs says we should give 2 characters per byte to display binary data in hex form */
     }
     else if (charset == NULL || charset->char_maxlen < 2/*i.e.0||1*/)
     {
-      return Field->length;
+      return MIN(Field->length, S2_1_GB);
     }
     else
     {
-      return Field->length/charset->char_maxlen;
+      return MIN(Field->length/charset->char_maxlen, S2_1_GB);
     }
   }
+  // no SingleStore type falls to the group below
+  case MYSQL_TYPE_ENUM:
+  case MYSQL_TYPE_SET:
+  case MYSQL_TYPE_GEOMETRY:
+  case MYSQL_TYPE_VARCHAR:
```

Comment:
Don't we have varchar?

Reply:
we do, but it is reported as MYSQL_TYPE_VAR_STRING.